hudi-common/src/main/java/org/apache/hudi/avro/JoinedGenericRecord.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.avro;
+
+import org.apache.hudi.common.model.HoodieRecord;
+
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericRecord;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Impl of {@link GenericRecord} to abstract meta fields and an actual data records of type GenericRecord.
+ */
+public class JoinedGenericRecord implements GenericRecord {
+  private final GenericRecord dataRecord;
+  private final Map<String, Object> metaFields;
+  private final List<String> metaFieldNames;
```

Comment:
There is no need to keep the `metaFields` as a map here, just an array should be enough because it is sequencial and also fixed size.